lager/lenses.hpp

```diff
@@ -123,51 +128,103 @@ auto attr(Member member)
+auto at(Key key) {
+    return zug::comp([key](auto&& f) {
+        return [f, &key](auto&& whole) {
+            using Part = std::optional<std::decay_t<decltype(whole.at(key))>>;
+
+            return f([&]() -> Part {
                 try {
-                    return std::forward<decltype(p)>(p).at(key);
-                } catch (std::out_of_range const&) {
-                    return std::decay_t<decltype(p.at(key))>{};
+                    return std::forward<decltype(whole)>(whole).at(key);
+                } catch (std::out_of_range const&) { return std::nullopt; }
+            }())([&](Part part) {
+                auto r = std::forward<decltype(whole)>(whole);
+                if (part.has_value()) {
+                    try {
+                        r.at(key) = std::move(part).value();
+                    } catch (std::out_of_range const&) {}
                 }
-            }())([&](auto&& x) {
-                auto r = std::forward<decltype(p)>(p);
-                try {
-                    r.at(key) = std::forward<decltype(x)>(x);
-                } catch (std::out_of_range const&) {}
                 return r;
             });
         };
     });
 }
 
+/*!
+ * `Key -> Lens<{X}, [X]>`
+ */
 template <typename Key>
-auto at_i(Key key)
-{
-    return zug::comp([=](auto&& f) {
-        return [f, &key](auto&& p) {
-            return f([&] {
+auto at_i(Key key) {
+    return zug::comp([key](auto&& f) {
+        return [f, &key](auto&& whole) {
+            using Part = std::optional<std::decay_t<decltype(whole.at(key))>>;
+
+            return f([&]() -> Part {
                 try {
-                    return std::forward<decltype(p)>(p).at(key);
-                } catch (std::out_of_range const&) {
-                    return std::decay_t<decltype(p.at(key))>{};
-                }
-            }())([&](auto&& x) {
-                if (static_cast<std::size_t>(key) < p.size()) {
-                    return std::forward<decltype(p)>(p).set(
-                        key, std::forward<decltype(x)>(x));
+                    return std::forward<decltype(whole)>(whole).at(key);
+                } catch (std::out_of_range const&) { return std::nullopt; }
+            }())([&](Part part) {
+                if (part.has_value() &&
+                    static_cast<std::size_t>(key) < whole.size()) {
+                    return std::forward<decltype(whole)>(whole).set(
+                        key, std::move(part).value());
                 } else {
-                    return std::forward<decltype(p)>(p);
+                    return std::forward<decltype(whole)>(whole);
                 }
             });
         };
     });
 }
 
+/*!
+ * `X -> Lens<[X], X>`
+ */
+template <typename T>
+auto fallback(T&& t) {
+    return zug::comp([t = std::forward<T>(t)](auto&& f) {
+        return [&, f](auto&& whole) {
+            return f(LAGER_FWD(whole).value_or(std::move(t)))(
+                [&](auto&& x) { return LAGER_FWD(x); });
```

Comment:
see [this comment](https://gist.github.com/Siapran/6e6a87ac5043e13a89bbb289dad4aacf#file-overkill-hpp-L116)